python/src/scipp/coords.py

```diff
@@ -0,0 +1,151 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
```

Comment:
The user defined graphs can have cycles but that would be invalid. We should check for that if it is not too difficult. Might have to either
- check the graph ahead of time, this would require inspecting all function signatures even of functions that are unused
- do it in line, this would require keeping track of nodes that have been visited already, maybe introduces needless complication in the main function.